dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -0,0 +1,784 @@
+        final String lockName = lockAddress.toString();
+        return List.of(EventFactory.Pthread.newLock(lockName, lockAddress));
+    }
+
+    private List<Event> inlinePthreadMutexUnlock(FunctionCall call) {
+        checkArgument(call.getArguments().size() == 1);
+        final Expression lockAddress = call.getArguments().get(0);
+        final String lockName = lockAddress.toString();
+        return List.of(EventFactory.Pthread.newUnlock(lockName, lockAddress));
+    }
+
+    private List<Event> inlineMalloc(FunctionCall call) {
+        if (call.getArguments().size() != 1) {
+            throw new UnsupportedOperationException(String.format("Unsupported signature for %s.", call));
+        }
+        final ValueFunctionCall valueCall = (ValueFunctionCall) call;
+        return List.of(EventFactory.newAlloc(
+                valueCall.getResultRegister(),
+                TypeFactory.getInstance().getByteType(),
+                valueCall.getArguments().get(0),
+                true
+        ));
+    }
+
+    private List<Event> inlineAssert(FunctionCall call) {
+        ExpressionFactory expressions = ExpressionFactory.getInstance();
+        final Expression condition = expressions.makeFalse();
+        final Event assertion = EventFactory.newAssert(condition, "user assertion");
+        final Event abort = EventFactory.newAbortIf(expressions.makeTrue());
+        abort.addTags(Tag.EARLYTERMINATION);
+        return List.of(assertion, abort);
+    }
+
+    // --------------------------------------------------------------------------------------------------------
+    // LLVM intrinsics
+
+    private List<Event> handleLLVMIntrinsic(FunctionCall call) {
+        assert call instanceof ValueFunctionCall && call.isDirectCall();
+        final ValueFunctionCall valueCall = (ValueFunctionCall) call;
+        final String name = call.getCalledFunction().getName();
+
+        if (name.startsWith("llvm.assume")) {
+            return List.of(EventFactory.newAssume(call.getArguments().get(0)));
+        } else if (name.startsWith("llvm.ctlz")) {
+            return inlineLLVMCtlz(valueCall);
+        } else if (name.startsWith("llvm.ctpop")) {
+            return inlineLLVMCtpop(valueCall);
+        } else if (name.contains("add.sat")) {
+            return inlineLLVMSaturatedAdd(valueCall);
+        } else if (name.contains("sub.sat")) {
+            return inlineLLVMSaturatedSub(valueCall);
+        } else if (name.startsWith("llvm.smax") || name.startsWith("llvm.smin")
+                || name.startsWith("llvm.umax") || name.startsWith("llvm.umin")) {
+            return inlineLLVMMinMax(valueCall);
+        } else {
+            final String error = String.format(
+                    "Call %s to LLVM intrinsic %s cannot be handled.", call, call.getCalledFunction());
+            throw new UnsupportedOperationException(error);
+        }
+    }
+
+    private List<Event> inlineLLVMCtlz(ValueFunctionCall call) {
```

Comment:
don't we have an expression to deal with?

Reply:
The expression might still be there, but I wanted to get rid of it since we can just implement it directly in our IR.
It would be strange to have `ctlz` as expression but not the other tons of llvm intrinsics. Similarly, I wouldn't want to add expressions for all the llvm intrinsics (if we don't need to)

Reply:
This class generates a long data dependency plus control flow which, as stated by a comment, it introduces dependencies that should not be there. The former might have performance impact wrt the one explicitly encoding `ctlz` semantics at the SMT level. The former impacts soundness of our analysis.

I think those two are sufficient reasons to treat this particular intrinsic differently than the rest.

Reply:
I agree with the ctrl-dep but that should get fixed by allowing jumps that do not produce dependencies.
The performance argument I only buy if you find a measurable difference in some code (even synthetic code would be fine).
Handling intrinsic `ctlz` expressions is not just a matter of adding some encoding in the backend, but you also need to provide handling in all processing, particular in CP. We get CP for free by giving an implementation.
Btw. ctpop suffers from the same long dependency chains.

Reply:
ok, but then the ctrl-dep problem should be fixed as part of this PR

Reply:
Then I would suggest to use the `ctlz` expression for now (the code is still there, just commented out) and make the changes in a latter PR.
This PR is already big enough as it is.

Reply:
We are still using the implementation (which introduces ctrl dependencies) rather than our own expression

Reply:
Not anymore.  Calls to `llvm.ctlz` now produce our built-in `CTLZ` expression.

Reply:
How? The code appears to be unchanged

Reply:
This diff viewer appears to be outdated.  You can look at the [actual code of the branch](https://github.com/hernanponcedeleon/Dat3M/blob/b2d17f1eca6a0a88d5aa00369acb4bc78294572f/dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java#L402).

Reply:
I think this happened because the marked code of this conversation, the method header, has not changed.